IPFS (#71)
eurec4a/how_to_eurec4a: how_to_eurec4a/data_on_ipfs.md

  language: python
  name: python3
---

# Data on IPFS

We use [IPFS](https://ipfs.io/) to store public data from the field campaign in a decentralized manner in order to increase data availability and accessibility.
Ideally, as a user of the How to EUREC4A book, you'd maybe not even notice that this system is working, because [intake](https://intake.readthedocs.io/en/latest/) handles the details of the data retrieval.

## Access via intake
Currently, the use of IPFS within the How to EUREC4A book is mostly optional and can be enabled by the `use_ipfs` argument of `eurec4a.get_intake_catalog` (it's disabled by default):

```{code-cell} ipython3
import eurec4a
cat = eurec4a.get_intake_catalog(use_ipfs=False)
list(sorted(cat))[:5]
```
This above command returns the first five entries in our [intake catalog](https://github.com/eurec4a/eurec4a-intake/#readme). Most of the data references by this catalog is accessed via the [OPeNDAP](https://www.opendap.org/) protocol from several servers in use by the EUREC4A community.
Let's look how this changes if we enable `use_ipfs`:

```{code-cell} ipython3
cat = eurec4a.get_intake_catalog(use_ipfs=True)
list(sorted(cat))[:5]
```

> ... we get the same result 😃. Ideally, the resulting intake catalog should behave just as the "classical" counterpart. Apart from a few edge-cases, which often are related to {doc}`netcdf_datatypes`, you should be able use both catalogs interchangeably, although almost everything behind the scenes works much differently.

Author: The IPFS catalog follows the non-IPFS catalog, so it should contain the same data. However, as the idea of the IPFS based method is to make data more available, it isn't identical (mainly because it shouldn't reproduce identical errors). I tried to explain this a bit in the text.

* They should be **easy to use** (it's important that people *want* to use these references, such that dataset creators use datasets in their published form).

Many popular references only provide some of these goals, e.g.:
* *filesystem paths* are not globally available and generally not immutable
* *HTTP-links* are global, but often less performant than local files. They are also not immutable and not by themselves fault tolerant
* *DOIs* don't refer directly to the data, as they are mostly about metadata. In most cases a DOI just forwards to some HTTP link, providing only some contractual guarantees on immutability and availability, but little technical benefits.

There's one unfortunate piece, which all of these references have in common: they always include (implicit or explicit) a server or computing system which is responsible to host the data.
They are references to the **location** of the data, not the **identity** of the data.

> This is much like describing the location of a book in a bookshelf (and the location of the bookshelf...) in stead of describing the book's identity by ISBN or author / title.

Author: Thanks :-) But the credit probably should go to [proto.school](https://proto.school/content-addressing/01).

Let's try it out:

```{code-cell} ipython3
store = CAStore()
k_hello = store.put(b"hello")
k_Hello = store.put(store.get(k_hello).replace(b"h", b"H"))
```

As we know the key of `hello` from before, we can now use that to ask our store for that value:

```{code-cell} ipython3
store.get('2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824')
```

Note that the modified version of `hello` with a capital `H` is also in the store, but of course under a different key:

```{code-cell} ipython3
store.get(k_Hello)
```

Finding out the value of `k_Hello` is left as an exercise to the reader 😃.

Another interesting feature is automatic de-duplication.
Let's see what happens to the amount of data stored before and after putting `hello` again:

```{code-cell} ipython3
print(f"item count before put: {len(store._store)}")
store.put(b"hello")
print(f"item count after put: {len(store._store)}")
```

Because every item is stored under it's own hash, identical items will only be stored once.

So a content addressable store allows to `put` and `get` some content.
Content which is already in the store won't change over time.

> If two different stores contain the same content, that same content will be retrievable under the same identifier, independent on which machine the store is.

Reviewer: ```suggestion
If two different stores contain the same content, that same content will be retrievable under the same identifier, independent of the machine hosting the store.
```

So far, we've got a local, **immutable** store, which gives **direct** access to the content stored.

> As it is a local store, access is as **performant** as any other local key-based access method would be (with automatic de-duplication as potential advantage in terms of required storage).

Reviewer: So `local` is good, but why again is it `local` in this case?

Author: Up to that point, the store is local because we didn't yet implement any communication. A local store is fast, but only for the data which is available...
I tried to explain this a bit more thoroughly.

For this purpose, IPFS builds a [distributed hash table](https://en.wikipedia.org/wiki/Distributed_hash_table) (DHT) as common in other peer-to-peer networks across all participating nodes.
The DHT contains a mapping from CID to node addresses.
Using this DHT, it is possible to find more peers which have the requested content stored without the need of any central registry.
Once connected to one of those newly discovered nodes, content can be retrieved from those nodes as above.
Because the DHT enables a **global** search for nodes which have the content behind a CID stored, it indeed becomes possible to use CIDs as a global identifier.

Whenever a node retrieves data behind a CID, it may keep a copy of that data for a while in a local cache.
Subsequent requests to the same CID can then be fulfilled locally.
Thus if a node is running on the computer of the user of a dataset, a first request to a dataset identified by CID will download the data, while future requests will operate on a local copy, leading to good access **performance** as mentioned in the beginning.

### ease of use
Ease of use is probably the most difficult to quantify.

> For read access to the data, IPFS should be very similar to what we know.

Reviewer: `...similar to what we know` Do you mean something like a normal file system?

Author: Yes, that's what I meant. I added some more words on this.

> In order to access datasets on the EUREC4A intake catalog via IPFS, you'll need to have `ipfsspec` installed in addition to what's needed anyways:

Reviewer: ```suggestion
In order to access datasets on the EUREC4A intake catalog via IPFS, you'll need to have `ipfsspec` installed in addition to your usual python packages:
```

```
pip install ipfsspec
``` 

> Afterwards, you can use links of the form `ipfs://<CID>` or `ipfs://<CID>/some/path/within/the/content` wherever `fsspec` is used to access files.

Reviewer: Could you add another sentence on `fsspec`?

Author: I've added an introduction, explaining that `fsspec` is used by `zarr` and `intake`.
